File: weird_captcha_gym/shared_runtime/app/mechanics/one_stroke_atelier.js
(() => {
  "use strict";
  let model = null;
  const clean = value => String(value == null ? "" : value).replaceAll("&", "&amp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;").replaceAll('"', "&quot;");
  const distance = (a, b) => Math.hypot(a[0] - b[0], a[1] - b[1]);
  const arrows = {right: "→", left: "←", up: "↑", down: "↓"};
  const labels = {tool: "TOOL", colour: "COLOUR", width: "WIDTH", stamp: "STAMP", burnish: "BURNISH"};
  const key = () => `${model.phase}|${model.prefix.join("/")}`;

  function record(kind, details = {}) {
    const sources = model.interaction === "full"
      ? {stroke_start: "direct_stroke", gate_cross: "direct_stroke", motif_sample: "direct_stroke", stroke_end: "direct_stroke"}
      : {stroke_start: "proxy_stroke", gate_cross: "proxy_gate", motif_sample: "proxy_motif", stroke_end: "proxy_stroke"};
    model.events.push({sequence: model.events.length + 1, kind, input_source: sources[kind], ...details});
  }
  function stagePoint(event) {
    const stage = document.querySelector(".atelier-stage"), rect = stage.getBoundingClientRect();
    return [
      Math.max(0, Math.min(model.state.stage.width, Math.round((event.clientX - rect.left - stage.clientLeft) / stage.clientWidth * model.state.stage.width))),
      Math.max(0, Math.min(model.state.stage.height, Math.round((event.clientY - rect.top - stage.clientTop) / stage.clientHeight * model.state.stage.height))),
    ];
  }
  function pointSegmentDistance(point, first, second) {
    const vx = second[0] - first[0], vy = second[1] - first[1], lengthSq = vx * vx + vy * vy;
    const t = lengthSq ? Math.max(0, Math.min(1, ((point[0] - first[0]) * vx + (point[1] - first[1]) * vy) / lengthSq)) : 0;
    const projection = [first[0] + vx * t, first[1] + vy * t];
    return {distance: distance(point, projection), projection};
  }
  function crossing(first, second, gate) {
    const [x, y] = gate.center, half = gate.hit_half_length ?? gate.half_length + gate.tolerance;
    if (gate.orientation === "vertical") {
      const delta = second[0] - first[0];
      if (!delta || (gate.direction === "right" && delta <= 0) || (gate.direction === "left" && delta >= 0)) return null;
      if (x < Math.min(first[0], second[0]) || x > Math.max(first[0], second[0])) return null;
      const t = (x - first[0]) / delta, iy = first[1] + t * (second[1] - first[1]);
      return Math.abs(iy - y) <= half ? t : null;
    }
    const delta = second[1] - first[1];
    if (!delta || (gate.direction === "down" && delta <= 0) || (gate.direction === "up" && delta >= 0)) return null;
    if (y < Math.min(first[1], second[1]) || y > Math.max(first[1], second[1])) return null;
    const t = (y - first[1]) / delta, ix = first[0] + t * (second[0] - first[0]);
    return Math.abs(ix - x) <= half ? t : null;
  }
  function hitsLockedBar(first, second, gate) {
    const [x, y] = gate.center, along = gate.hit_half_length ?? gate.half_length + gate.tolerance;
    const halfX = gate.orientation === "vertical" ? 9 : along;
    const halfY = gate.orientation === "vertical" ? along : 9;
    const bounds = [[x - halfX, x + halfX], [y - halfY, y + halfY]];
    const delta = [second[0] - first[0], second[1] - first[1]];
    let low = 0, high = 1;
    for (let axis = 0; axis < 2; axis += 1) {
      if (Math.abs(delta[axis]) < 1e-9) {
        if (first[axis] < bounds[axis][0] || first[axis] > bounds[axis][1]) return false;
        continue;
      }
      let enter = (bounds[axis][0] - first[axis]) / delta[axis];
      let leave = (bounds[axis][1] - first[axis]) / delta[axis];
      if (enter > leave) [enter, leave] = [leave, enter];
      low = Math.max(low, enter); high = Math.min(high, leave);
      if (low > high) return false;
    }
    return true;
  }
  function currentGates() { return model.state.gate_sets[key()] || []; }
  function appendInk(point) {
    model.ink.push(point);
    const polyline = document.querySelector(".atelier-ink");
    if (polyline) polyline.setAttribute("points", model.ink.map(item => item.join(",")).join(" "));
  }
  function updateReceipt() {
    document.querySelectorAll(".atelier-receipt-row").forEach(row => {
      const field = row.dataset.field, value = model.selected[field];
      row.dataset.filled = value ? "true" : "false";
      const output = row.querySelector("b");
      if (output) output.textContent = value ? value.toUpperCase() : "—";
    });
    const count = document.querySelector(".atelier-stroke-count");
    if (count) count.textContent = `${model.strokeCount}/${model.state.stroke_budget}`;
  }
  function gateMarkup(gate, locked = false) {
    const along = gate.hit_half_length ?? gate.half_length + gate.tolerance;
    const width = gate.orientation === "horizontal" ? along * 2 : 18;
    const height = gate.orientation === "vertical" ? along * 2 : 18;
    return `<div class="atelier-gate atelier-gate--${gate.orientation}${locked ? " atelier-gate--locked" : ""}" data-gate-id="${clean(gate.id)}" data-direction="${gate.direction}" data-locked="${locked}" style="left:${gate.center[0] / model.state.stage.width * 100}%;top:${gate.center[1] / model.state.stage.height * 100}%;width:${width / model.state.stage.width * 100}%;height:${height / model.state.stage.height * 100}%;--swatch:${clean(gate.swatch || "#d1ab58")}"><i></i><span class="atelier-glyph">${clean(gate.glyph)}</span><strong>${clean(gate.label)}</strong><em>${locked ? "×" : arrows[gate.direction]}</em></div>`;
  }
  function motifMarkup() {
    const points = model.state.motif.points;
    return `<svg class="atelier-jig" viewBox="0 0 ${model.state.stage.width} ${model.state.stage.height}" aria-label="badge tracing jig"><polyline points="${points.map(item => item.join(",")).join(" ")}"/></svg>${points.map((point, index) => `<button type="button" class="atelier-checkpoint" data-index="${index}" data-active="${index === model.motifIndex}" style="left:${point[0] / model.state.stage.width * 100}%;top:${point[1] / model.state.stage.height * 100}%"><span>${index === points.length - 1 ? "FINISH" : index + 1}</span></button>`).join("")}`;
  }
  function bindDynamic() {
    return;
  }
  function renderDynamic() {
    const layer = document.querySelector(".atelier-dynamic");
    if (!layer) return;
    const locked = `<div class="atelier-locked-layer">${model.lockedGates.map(gate => gateMarkup(gate, true)).join("")}</div>`;
    if (model.phase < model.state.active_fields.length) {
      const field = model.state.active_fields[model.phase];
      layer.innerHTML = `${locked}<div class="atelier-bank-title"><b>${labels[field]}</b></div>${currentGates().map(gate => gateMarkup(gate)).join("")}`;
    } else {
      layer.innerHTML = `${locked}<div class="atelier-bank-title atelier-bank-title--jig"><b>MOTIF</b></div>${motifMarkup()}`;
    }
    bindDynamic();
    updateReceipt();
  }
  function activateGate(gate, before, after, pathSegment = null) {
    const field = model.state.active_fields[model.phase];
    record("gate_cross", {stroke: model.strokeCount, gate_id: gate.id, field, value: gate.value, direction: gate.direction, before, after, path_segment: pathSegment});
    model.selected[field] = gate.value;
    model.prefix.push(gate.value);
    if (model.state.locked_gate_memory > 0) {
      model.lockedGates.push({...gate});
      model.lockedGates = model.lockedGates.slice(-model.state.locked_gate_memory);
    }
    model.phase += 1;
    renderDynamic();
    model.helpers.setReadout(clean(gate.label), "pending");
  }
  function reachMotif(point, pathSegment = null) {
    const index = model.motifIndex;
    if (!model.motifTrace.length || distance(model.motifTrace.at(-1), point) > 1) model.motifTrace.push(point);
    record("motif_sample", {stroke: model.strokeCount, checkpoint: index, point, path_segment: pathSegment});
    model.motifIndex += 1;
    renderDynamic();
    model.helpers.setReadout("MOTIF", "pending");
  }
  function moveStroke(point) {
    if (!model.inStroke) return;
    const before = model.lastPoint;
    appendInk(point);
    const pathSegment = model.ink.length - 2;
    const lockedHit = model.lockedGates.find(gate => hitsLockedBar(before, point, gate));
    if (model.invalid || lockedHit) {
      if (lockedHit && !model.invalid) model.routeViolations.push({kind: "locked_bar", gate_id: lockedHit.id, before, after: point});
      model.invalid = true;
      model.lastPoint = point;
      document.querySelector(".atelier-stage")?.setAttribute("data-invalid", "true");
      model.helpers.setReadout("FAIL", "error");
      return;
    }
    if (model.phase < model.state.active_fields.length) {
      const hits = currentGates().map(gate => ({gate, t: crossing(before, point, gate)})).filter(item => item.t != null).sort((a, b) => a.t - b.t);
      if (hits.length) activateGate(hits[0].gate, before, point, pathSegment);
    } else if (model.motifIndex < model.state.motif.points.length) {
      const expected = model.state.motif.points[model.motifIndex], hit = pointSegmentDistance(expected, before, point);
      if (model.motifIndex > 0) model.motifTrace.push(point);
      if (hit.distance <= model.state.motif.tolerance) reachMotif(hit.projection.map(Math.round), pathSegment);
    }
    model.lastPoint = point;
  }
  function startStroke(point, event = null) {
    if (model.inStroke || model.completed || model.submitting || model.strokeCount >= model.state.stroke_budget) return;
    if (distance(point, model.state.start) > 38) return;
    model.strokeCount += 1;
    model.inStroke = true;
    model.lastPoint = point;
    model.phase = 0;
    model.prefix = [];
    model.selected = {};
    model.motifIndex = 0;
    model.motifTrace = [];
    model.lockedGates = [];
    model.ink = [point];
    model.events = [];
    model.routeViolations = [];
    model.invalid = false;
    record("stroke_start", {stroke: model.strokeCount, point, path_index: 0});
    if (event) {
      event.preventDefault();
      model.activePointerId = event.pointerId;
      document.querySelector(".atelier-stage").setPointerCapture(event.pointerId);
    }
    document.querySelector(".atelier-stage")?.setAttribute("data-stroking", "true");
    document.querySelector(".atelier-stage")?.setAttribute("data-invalid", "false");
    renderDynamic();
    model.helpers.setReadout("STROKE", "pending");
  }
  function resetForRetry() {
    model.phase = 0; model.prefix = []; model.selected = {}; model.motifIndex = 0; model.motifTrace = []; model.lockedGates = []; model.ink = []; model.events = []; model.routeViolations = []; model.invalid = false; model.completed = false;
    document.querySelector(".atelier-ink")?.setAttribute("points", "");
    document.querySelector(".atelier-stage")?.setAttribute("data-invalid", "false");
    document.querySelector(".atelier-shell")?.setAttribute("data-complete", "false");
    const submitButton = document.querySelector(".atelier-submit");
    if (submitButton) submitButton.disabled = true;
    renderDynamic();
  }
  function cancelStroke(termination) {
    if (!model.inStroke) return;
    model.interruptions.push({sequence: model.interruptions.length + 1, kind: "stroke_cancel", input_source: "direct_stroke", termination, stroke: model.strokeCount, complete: false});
    model.inStroke = false;
    model.activePointerId = null;
    model.strokeCount = Math.max(0, model.strokeCount - 1);
    document.querySelector(".atelier-stage")?.setAttribute("data-stroking", "false");
    resetForRetry();
    model.helpers.setReadout("FAIL", "error");
    updateReceipt();
  }
  function endStroke(point, termination) {
    if (!model.inStroke) return;
    if (model.interaction === "full" && distance(model.lastPoint, point) > 0) {
      appendInk(point);
      model.lastPoint = point;
    }
    const last = model.state.motif.points.at(-1), complete = !model.invalid && model.motifIndex === model.state.motif.points.length && distance(point, last) <= model.state.motif.tolerance;
    record("stroke_end", {stroke: model.strokeCount, point, complete, termination, path_index: model.ink.length - 1});
    model.inStroke = false;
    model.activePointerId = null;
    document.querySelector(".atelier-stage")?.setAttribute("data-stroking", "false");
    if (complete) {
      model.completed = true;
      document.querySelector(".atelier-shell")?.setAttribute("data-complete", "true");
      document.querySelector(".atelier-submit").disabled = false;
      model.helpers.setReadout("READY", "passed");
    } else if (model.strokeCount < model.state.stroke_budget) {
      resetForRetry();
      model.helpers.setReadout("FAIL", "error");
    } else {
      document.querySelector(".atelier-submit").disabled = false;
      model.helpers.setReadout("FAIL", "error");
    }
    updateReceipt();
  }
  function proxyStart() { startStroke(model.state.start); }
  function proxyEnd() {
    endStroke(model.lastPoint || model.state.start, "proxy_end");
  }
  async function submit() {
    if (model.submitting || model.terminal || (!model.completed && model.strokeCount < model.state.stroke_budget)) return;
    model.submitting = true;
    model.helpers.setReadout("READY", "pending");
    const payload = {mechanic_id: model.state.mechanic_id, challenge_id: model.state.challenge_id, interaction: model.interaction, events: model.events, interruptions: model.interruptions, route_violations: model.routeViolations, selected_fields: model.selected, drawn_geometry: model.motifTrace, stroke_geometry: model.ink, stroke_count: model.strokeCount, completed: model.completed};
    try {
      const response = await fetch("/result", {method: "POST", headers: {"content-type": "application/json"}, body: JSON.stringify(payload)});
      if (!response.ok) throw new Error(`HTTP ${response.status}`);
      const outcome = await response.json();
      if (outcome.passed === true) {
        model.terminal = true;
        document.querySelector(".atelier-shell")?.insertAdjacentHTML("beforeend", '<div class="atelier-verdict"><strong>PASS</strong></div>');
        model.helpers.setReadout("PASS", "passed");
      } else if (outcome.passed === false && outcome.state) {
        await model.helpers.render(outcome.state);
        const shell = document.querySelector(".atelier-shell");
        shell?.setAttribute("data-fresh-failure", "true");
        shell?.insertAdjacentHTML("afterbegin", '<div class="atelier-failure"><b>FAIL</b></div>');
        model.helpers.setReadout("FAIL", "error");
      } else throw new Error("authoritative grade unavailable");
    } catch (_error) {
      model.submitting = false;
      document.querySelector(".atelier-submit").disabled = false;
      model.helpers.setReadout("READY", "idle");
    }
  }
  function targetMarkup(item) {
    return `<div class="atelier-target-field"><span>${labels[item.field]}</span><b>${clean(item.glyph)} ${clean(item.label)}</b>${item.swatch ? `<i style="background:${clean(item.swatch)}"></i>` : ""}</div>`;
  }
  async function render(state, helpers) {
    document.body.dataset.mechanic = "one-stroke-atelier";
    document.body.dataset.cheatMode = helpers.isCheatMode() ? "true" : "false";
    const interaction = state.control_condition?.interaction || "full", motif = state.motif.points;
    model = {state, helpers, interaction, events: [], interruptions: [], routeViolations: [], phase: 0, prefix: [], selected: {}, motifIndex: 0, motifTrace: [], lockedGates: [], ink: [], lastPoint: null, activePointerId: null, inStroke: false, invalid: false, strokeCount: 0, completed: false, submitting: false, terminal: false};
    window.oneStrokeAtelierModel = model;
    helpers.app.innerHTML = `<section class="atelier-shell" data-challenge-id="${clean(state.challenge_id)}" data-interaction="${interaction}" data-complete="false"><header class="atelier-header"><div><span>MAISON VERVAIN · ENAMEL ROOM</span><h1>One-Stroke Atelier</h1><p>${clean(state.prompt)}</p></div><div class="atelier-commission"><span>COMMISSION</span><b>No. ${clean(state.challenge_id)}</b><small>${interaction === "full" ? "DIRECT LATHE" : "PROXY CONSOLE"}</small></div></header><main class="atelier-main"><aside class="atelier-target"><div class="atelier-target-head"><span>TARGET</span></div><div class="atelier-badge"><svg viewBox="580 200 310 150"><polyline points="${motif.map(item => item.join(",")).join(" ")}"/></svg><i>${clean(state.target.find(item => item.field === "stamp")?.glyph || "✦")}</i></div><div class="atelier-target-fields">${state.target.map(targetMarkup).join("")}</div></aside><section class="atelier-bench"><div class="atelier-stage" data-stroking="false" data-invalid="false"><div class="atelier-grain"></div><svg class="atelier-ink-svg" viewBox="0 0 ${state.stage.width} ${state.stage.height}"><polyline class="atelier-ink" points=""/></svg><div class="atelier-start" style="left:${state.start[0] / state.stage.width * 100}%;top:${state.start[1] / state.stage.height * 100}%"><i></i><b>START</b><span>${interaction === "full" ? "HOLD" : "PROXY"}</span></div><div class="atelier-dynamic"></div></div><div class="atelier-controls">${interaction === "simplified" ? '<button type="button" class="atelier-proxy-start">BEGIN</button><button type="button" class="atelier-proxy-end">END</button>' : ''}<div class="atelier-readout"><span>STATUS</span><div class="readout" data-status="idle">READY</div></div></div></section><aside class="atelier-ledger"><div class="atelier-ledger-head"><span>RECEIPT</span><b>STROKE <i class="atelier-stroke-count">0/${state.stroke_budget}</i></b></div>${state.active_fields.map(field => `<div class="atelier-receipt-row" data-field="${field}" data-filled="false"><span>${labels[field]}</span><b>—</b></div>`).join("")}<button type="button" class="atelier-submit" disabled>${clean(state.submit_label)}</button></aside></main>${helpers.cheatPanelTemplate()}</section>`;
    const stage = document.querySelector(".atelier-stage");
    if (interaction === "full") {
      stage.addEventListener("pointerdown", event => { if (event.button === 0) startStroke(stagePoint(event), event); });
      stage.addEventListener("pointermove", event => moveStroke(stagePoint(event)));
      stage.addEventListener("pointerup", event => endStroke(stagePoint(event), "pointerup"));
      stage.addEventListener("pointercancel", () => cancelStroke("pointercancel"));
      stage.addEventListener("lostpointercapture", () => cancelStroke("lostpointercapture"));
    } else {
      document.querySelector(".atelier-proxy-start").addEventListener("click", proxyStart);
      document.querySelector(".atelier-proxy-end").addEventListener("click", proxyEnd);
      stage.addEventListener("click", event => moveStroke(stagePoint(event)));
    }
    document.querySelector(".atelier-submit").addEventListener("click", submit);
    renderDynamic();
    helpers.installCheatPanel();
  }
  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};
  window.WeirdCaptchaMechanics.one_stroke_atelier = {rootSelector: ".atelier-shell", render};
})();

File: weird_captcha_gym/tools/incubator_solvers/one_stroke_atelier.py
from __future__ import annotations

import json
import heapq
import math
import time
from pathlib import Path

from playwright.sync_api import expect


MECHANIC_ID = "one_stroke_atelier"


def _read(path: Path) -> dict:
    return json.loads(path.read_text(encoding="utf-8"))


def _box(page) -> dict:
    stage = page.locator(".atelier-stage")
    value = stage.bounding_box()
    if not value:
        raise AssertionError("atelier stage is not visible")
    client = stage.evaluate("element => ({left: element.clientLeft, top: element.clientTop, width: element.clientWidth, height: element.clientHeight})")
    return {"x": value["x"] + client["left"], "y": value["y"] + client["top"], "width": client["width"], "height": client["height"]}


def _screen(box: dict, stage: dict, point: list[float]) -> tuple[float, float]:
    return box["x"] + point[0] / stage["width"] * box["width"], box["y"] + point[1] / stage["height"] * box["height"]


def _gate_key(phase: int, prefix: list[str]) -> str:
    return f"{phase}|{'/'.join(prefix)}"


def _target_map(truth: dict) -> dict[str, str]:
    return {item["field"]: item["value"] for item in truth["target"]}


def _gate(truth: dict, phase: int, prefix: list[str], wrong: bool = False) -> dict:
    field = truth["active_fields"][phase]
    target = _target_map(truth)[field]
    gates = truth["gate_sets"][_gate_key(phase, prefix)]
    return next(item for item in gates if (item["value"] != target if wrong else item["value"] == target))


def _move(page, box: dict, stage: dict, point: list[float]) -> None:
    page.mouse.move(*_screen(box, stage, point), steps=3)


def _advance(page, box: dict, stage: dict, point: list[float], *, proxy: bool) -> None:
    if proxy:
        page.mouse.click(*_screen(box, stage, point))
    else:
        _move(page, box, stage, point)


def _hit_half_length(gate: dict) -> float:
    return float(gate.get("hit_half_length", gate["half_length"] + gate["tolerance"]))


def _segment_hits_bar(first: list[float], second: list[float], gate: dict) -> bool:
    x, y = (float(item) for item in gate["center"])
    along = _hit_half_length(gate)
    half_x, half_y = (9.0, along) if gate["orientation"] == "vertical" else (along, 9.0)
    bounds = ((x - half_x, x + half_x), (y - half_y, y + half_y))
    delta = (second[0] - first[0], second[1] - first[1])
    low, high = 0.0, 1.0
    for axis in range(2):
        if abs(delta[axis]) < 1e-9:
            if not bounds[axis][0] <= first[axis] <= bounds[axis][1]:
                return False
            continue
        enter = (bounds[axis][0] - first[axis]) / delta[axis]
        leave = (bounds[axis][1] - first[axis]) / delta[axis]
        if enter > leave:
            enter, leave = leave, enter
        low, high = max(low, enter), min(high, leave)
        if low > high:
            return False
    return True


def _route_around(start: list[float], goal: list[float], barriers: list[dict], stage: dict) -> list[list[float]]:
    if not any(_segment_hits_bar(start, goal, gate) for gate in barriers):
        return [list(start), list(goal)]
    inset, margin = 14.0, 13.0
    width, height = float(stage["width"]), float(stage["height"])
    candidates: list[list[float]] = [list(start), list(goal), [inset, inset], [width - inset, inset], [inset, height - inset], [width - inset, height - inset]]
    for gate in barriers:
        x, y = (float(item) for item in gate["center"])
        along = _hit_half_length(gate)
        if gate["orientation"] == "vertical":
            candidates.extend([[x - margin, y - along - margin], [x + margin, y - along - margin], [x - margin, y + along + margin], [x + margin, y + along + margin]])
        else:
            candidates.extend([[x - along - margin, y - margin], [x - along - margin, y + margin], [x + along + margin, y - margin], [x + along + margin, y + margin]])
    nodes: list[list[float]] = []
    seen: set[tuple[float, float]] = set()
    for index, point in enumerate(candidates):
        clamped = list(point) if index < 2 else [max(inset, min(width - inset, point[0])), max(inset, min(height - inset, point[1]))]
        token = (round(clamped[0], 3), round(clamped[1], 3))
        if token not in seen and not any(_segment_hits_bar(clamped, clamped, gate) for gate in barriers):
            seen.add(token)
            nodes.append(clamped)
    start_index, goal_index = nodes.index(list(start)), nodes.index(list(goal))
    graph: list[list[tuple[int, float]]] = [[] for _ in nodes]
    for first_index, first in enumerate(nodes):
        for second_index in range(first_index + 1, len(nodes)):
            second = nodes[second_index]
            if any(_segment_hits_bar(first, second, gate) for gate in barriers):
                continue
            weight = math.hypot(second[0] - first[0], second[1] - first[1])
            graph[first_index].append((second_index, weight))
            graph[second_index].append((first_index, weight))
    distances = [math.inf] * len(nodes)
    previous: list[int | None] = [None] * len(nodes)
    distances[start_index] = 0.0
    queue = [(0.0, start_index)]
    while queue:
        total, node = heapq.heappop(queue)
        if total != distances[node]:
            continue
        if node == goal_index:
            break
        for neighbor, weight in graph[node]:
            candidate = total + weight
            if candidate < distances[neighbor]:
                distances[neighbor], previous[neighbor] = candidate, node
                heapq.heappush(queue, (candidate, neighbor))
    if not math.isfinite(distances[goal_index]):
        raise AssertionError("generated atelier barrier layout has no visible route")
    route: list[list[float]] = []
    node: int | None = goal_index
    while node is not None:
        route.append(nodes[node])
        node = previous[node]
    return list(reversed(route))


def _cross_full(
    page,
    box: dict,
    stage: dict,
    current: list[float],
    gate: dict,
    current_gates: list[dict],
    locked_gates: list[dict],
    *,
    proxy: bool,
) -> list[float]:
    x, y = gate["center"]
    if gate["orientation"] == "vertical":
        before = [x - 34, y] if gate["direction"] == "right" else [x + 34, y]
        after = [x + 34, y] if gate["direction"] == "right" else [x - 34, y]
    else:
        before = [x, y - 34] if gate["direction"] == "down" else [x, y + 34]
        after = [x, y + 34] if gate["direction"] == "down" else [x, y - 34]
    for waypoint in _route_around(current, before, [*locked_gates, *current_gates], stage)[1:]:
        _advance(page, box, stage, waypoint, proxy=proxy)
    _advance(page, box, stage, after, proxy=proxy)
    return after


def _perform(page, truth: dict, *, wrong_first: bool = False, screenshot: Path | None = None) -> None:
    interaction = str((truth.get("control_condition") or {}).get("interaction") or "full")
    prefix: list[str] = []
    locked_gates: list[dict] = []
    box, stage = _box(page), truth["stage"]
    current = list(truth["start"])
    if interaction == "simplified":
        page.locator(".atelier-proxy-start").click()
    else:
        page.mouse.move(*_screen(box, stage, current))
        page.mouse.down()
    for phase in range(len(truth["active_fields"])):
        gate = _gate(truth, phase, prefix, wrong=wrong_first and phase == 0)
        gates = truth["gate_sets"][_gate_key(phase, prefix)]
        current = _cross_full(page, box, stage, current, gate, gates, locked_gates, proxy=interaction == "simplified")
        prefix.append(gate["value"])
        memory = int(truth.get("locked_gate_memory") or 0)
        if memory:
            locked_gates.append(gate)
            locked_gates = locked_gates[-memory:]
        page.wait_for_timeout(90)
        if screenshot and phase == 1:
            page.screenshot(path=str(screenshot))
    motif = truth["motif"]["points"]
    for waypoint in _route_around(current, motif[0], locked_gates, stage)[1:]:
        _advance(page, box, stage, waypoint, proxy=interaction == "simplified")
        current = list(waypoint)
    for point in motif[1:]:
        _advance(page, box, stage, point, proxy=interaction == "simplified")
        current = list(point)
        page.wait_for_timeout(70)
    if interaction == "simplified":
        page.locator(".atelier-proxy-end").click()
    else:
        page.mouse.up()


def solve(page, state_dir: Path, out_dir: Path, mechanic: str) -> None:
    if mechanic != MECHANIC_ID:
        raise AssertionError(f"unexpected mechanic {mechanic!r}")
    truth = _read(state_dir / "ground_truth.json")
    out_dir.mkdir(parents=True, exist_ok=True)
    _perform(page, truth, screenshot=out_dir / f"{mechanic}-mid-stroke-reconfiguration.png")
    expect(page.locator('.atelier-shell[data-complete="true"]')).to_be_visible()
    page.screenshot(path=str(out_dir / f"{mechanic}-completed-before-certify.png"))
    page.locator(".atelier-submit").click()
    expect(page.locator(".readout")).to_have_text("PASS", timeout=8_000)
    page.screenshot(path=str(out_dir / f"{mechanic}-pass.png"))


def fail_once(page, state_dir: Path, out_dir: Path, mechanic: str) -> None:
    if mechanic != MECHANIC_ID:
        raise AssertionError(f"unexpected mechanic {mechanic!r}")
    before = _read(state_dir / "ground_truth.json")["challenge_id"]
    _perform(page, _read(state_dir / "ground_truth.json"), wrong_first=True)
    page.locator(".atelier-submit").click()
    deadline = time.monotonic() + 8
    while time.monotonic() < deadline:
        current = _read(state_dir / "ground_truth.json")["challenge_id"]
        if current != before:
            break
        time.sleep(0.05)
    else:
        raise AssertionError("failed atelier submission did not generate a fresh challenge")
    expect(page.locator('.atelier-shell[data-fresh-failure="true"]')).to_be_visible()
    expect(page.locator(".readout")).to_contain_text("FAIL")
    out_dir.mkdir(parents=True, exist_ok=True)
    page.screenshot(path=str(out_dir / f"{mechanic}-fail-refresh.png"))
